packages/prop-house-backend/src/twitter/twitter.service.ts

```diff
@@ -0,0 +1,116 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { EventEmitter2 } from '@nestjs/event-emitter';
+import { InjectRepository } from '@nestjs/typeorm';
+import { PinataClient, PinataPinOptions } from '@pinata/sdk';
+import { AuctionClosedEvent } from 'src/auction/events/auction-closed.event';
+import { AuctionCreatedEvent } from 'src/auction/events/auction-created.event';
+import { AuctionOpenEvent } from 'src/auction/events/auction-open.event';
+import { AuctionProposalEndingSoonEvent } from 'src/auction/events/auction-proposal-end-soon.event';
+import { AuctionVotingEndingSoonEvent } from 'src/auction/events/auction-vote-end-soon.event';
+import { AuctionVotingEvent } from 'src/auction/events/auction-voting.event';
+import { EventStatus } from 'src/auction/types';
+import { ProposalCreatedEvent } from 'src/proposal/events/proposal-created.event';
+import {
+  SendTweetV2Params,
+  TweetV2PostTweetResult,
+  TwitterApi,
+} from 'twitter-api-v2';
+import { Repository } from 'typeorm';
+import { Tweet } from './tweet.entity';
+import { AuctionTweetable, Tweetable, TweetableContents } from './types';
+
+@Injectable()
+export class TwitterService {
+  private twitter: TwitterApi;
+  private readonly logger = new Logger(TwitterService.name);
+
+  constructor(
+    private readonly config: ConfigService,
+    private readonly events: EventEmitter2,
+    @InjectRepository(Tweet)
+    private readonly tweetRepository: Repository<Tweet>,
+  ) {
+    if (this.config.get<boolean>('social.twitter.enabled')) {
+      this.logger.verbose('Creating TwitterAPI Client');
+      this.twitter = new TwitterApi({
+        appKey: this.config.get<string>('social.twitter.appKey'),
+        appSecret: this.config.get<string>('social.twitter.appSecret'),
+        accessToken: this.config.get<string>('social.twitter.accessToken'),
+        accessSecret: this.config.get<string>('social.twitter.accessSecret'),
+      });
```

Comment:
nestjs nit: prefer construction in a custom provider, which is then injected into this service

Reply:
Pulled the client into its own provider :heavy_check_mark:

Reply:
Sorry, was referring to a custom provider that's created like:

```ts
providers: [
  TwitterClientService,
  TwitterService,
  {
    provide: TwitterProvider.TWITTER_API_CLIENT,
    useFactory: (config: ConfigService) => {
      return new TwitterApi({
        appKey: config.get<string>('social.twitter.appKey'),
        appSecret: config.get<string>('social.twitter.appSecret'),
        accessToken: config.get<string>('social.twitter.accessToken'),
        accessSecret: config.get<string>('social.twitter.accessSecret'),
      });
    },
    inject: [ConfigService],
  },
],
```

and then injected for usage wherever like:

```
constructor(
  @Inject(TwitterProvider.TWITTER_API_CLIENT)
  private readonly _twitter: TwitterApi,
) {}
```